File: alarm-manager/src/store.js
// 闹钟管家 - 本地存储模块（纯 Node API，可在主进程和测试中复用）
// 数据存储路径：app.getPath('userData')/alarms.json
// 设计目标：原子写入、UTF-8 编码、自动备份、导入导出

const fs = require('fs');
const path = require('path');

const DEFAULT_FILE = path.join(
  process.env.APPDATA || process.env.HOME || __dirname,
  'alarm-manager',
  'alarms.json'
);

// 默认数据结构
function defaultData() {
  return {
    version: 1,
    alarms: [],          // 闹钟列表
    settings: {
      defaultSound: 'chime',         // 默认铃声 key
      defaultSnoozeMinutes: 5,       // 默认贪睡分钟
      maxSnoozeCount: 3,             // 默认最大贪睡次数
      volumeFadeIn: true,            // 渐强
      volumeFadeInDuration: 15,      // 渐强秒数
      maxVolume: 0.9,                // 最大音量 0..1
      notificationEnabled: true,     // 系统通知
      bringToFront: true,            // 触发时窗口置顶
      autoStartAtLogin: false        // 开机自启
    },
    logs: []              // 触发历史（最近 200 条）
  };
}

// 合并默认值与读取值（避免新字段缺失）
function mergeWithDefaults(raw) {
  const base = defaultData();
  if (!raw || typeof raw !== 'object') return base;
  return {
    version: raw.version || base.version,
    alarms: Array.isArray(raw.alarms) ? raw.alarms : [],
    settings: Object.assign({}, base.settings, raw.settings || {}),
    logs: Array.isArray(raw.logs) ? raw.logs.slice(-200) : []
  };
}

function ensureDir(file) {
  const dir = path.dirname(file);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

// 读取数据，失败返回默认值
function load(file) {
  file = file || DEFAULT_FILE;
  try {
    if (!fs.existsSync(file)) return defaultData();
    const buf = fs.readFileSync(file);
    const text = buf.toString('utf8');
    const parsed = JSON.parse(text);
    return mergeWithDefaults(parsed);
  } catch (err) {
    return defaultData();
  }
}

// 原子写入：写到 .tmp，再重命名
function save(data, file) {
  file = file || DEFAULT_FILE;
  ensureDir(file);
  const text = JSON.stringify(data, null, 2);
  const tmp = file + '.tmp';
  fs.writeFileSync(tmp, text, 'utf8');
  // 备份上一版
  try {
    if (fs.existsSync(file)) {
      const bak = file + '.bak';
      fs.copyFileSync(file, bak);
    }
  } catch (e) { /* 备份失败不影响保存 */ }
  fs.renameSync(tmp, file);
  return true;
}

// 添加日志，保留最近 200 条
function appendLog(data, entry) {
  data.logs = Array.isArray(data.logs) ? data.logs : [];
  data.logs.push(Object.assign({ time: Date.now() }, entry));
  if (data.logs.length > 200) {
    data.logs = data.logs.slice(-200);
  }
}

// 导出 JSON（备份用）
function exportJson(data) {
  return JSON.stringify(data, null, 2);
}

// 导入 JSON（合并校验）
function importJson(text) {
  const parsed = JSON.parse(text);
  return mergeWithDefaults(parsed);
}

// 生成新闹钟 ID（短而唯一）
function newId() {
  return 'a_' + Date.now().toString(36) + '_' + Math.random().toString(36).slice(2, 8);
}

module.exports = {
  DEFAULT_FILE,
  defaultData,
  mergeWithDefaults,
  load,
  save,
  appendLog,
  exportJson,
  importJson,
  newId
};

File: alarm-manager/portable/Core/Store.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using AlarmManager.Portable.Models;

namespace AlarmManager.Portable.Core
{
    /// <summary>
    /// 本地存储模块 - 移植自 alarm-manager/src/store.js
    /// 设计目标：原子写入、UTF-8 编码、自动备份、导入导出
    /// </summary>
    public static class Store
    {
        private static readonly JsonSerializerOptions JsonOpts = new JsonSerializerOptions
        {
            WriteIndented = true,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.Never,
        };

        /// <summary>
        /// 默认数据文件：%APPDATA%/alarm-manager-portable/alarms.json
        /// </summary>
        public static string DefaultFile
        {
            get
            {
                string baseDir = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
                return Path.Combine(baseDir, "alarm-manager-portable", "alarms.json");
            }
        }

        public static AlarmData DefaultData() => new AlarmData();

        /// <summary>合并默认值与读取值（避免新字段缺失）</summary>
        public static AlarmData MergeWithDefaults(AlarmData? raw)
        {
            var baseData = DefaultData();
            if (raw == null) return baseData;
            return new AlarmData
            {
                Version = raw.Version == 0 ? baseData.Version : raw.Version,
                Alarms = raw.Alarms ?? new System.Collections.Generic.List<Alarm>(),
                Settings = MergeSettings(baseData.Settings, raw.Settings),
                Logs = (raw.Logs ?? new System.Collections.Generic.List<TriggerLog>())
                    .Skip(Math.Max(0, (raw.Logs?.Count ?? 0) - 200)).Take(200).ToList()
            };
        }

        private static AppSettings MergeSettings(AppSettings def, AppSettings? raw)
        {
            if (raw == null) return def;
            // 简单合并：raw 中存在的字段优先，缺失的用默认
            return new AppSettings
            {
                DefaultSound = string.IsNullOrEmpty(raw.DefaultSound) ? def.DefaultSound : raw.DefaultSound,
                DefaultSnoozeMinutes = raw.DefaultSnoozeMinutes == 0 ? def.DefaultSnoozeMinutes : raw.DefaultSnoozeMinutes,
                MaxSnoozeCount = raw.MaxSnoozeCount == 0 ? def.MaxSnoozeCount : raw.MaxSnoozeCount,
                VolumeFadeIn = raw.VolumeFadeIn,
                VolumeFadeInDuration = raw.VolumeFadeInDuration == 0 ? def.VolumeFadeInDuration : raw.VolumeFadeInDuration,
                MaxVolume = raw.MaxVolume <= 0 ? def.MaxVolume : raw.MaxVolume,
                NotificationEnabled = raw.NotificationEnabled,
                BringToFront = raw.BringToFront,
                AutoStartAtLogin = raw.AutoStartAtLogin,
            };
        }

        public static void EnsureDir(string file)
        {
            string dir = Path.GetDirectoryName(file);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }
        }

        /// <summary>读取数据，失败返回默认值</summary>
        public static AlarmData Load(string? file = null)
        {
            file = file ?? DefaultFile;
            try
            {
                if (!File.Exists(file)) return DefaultData();
                string text = File.ReadAllText(file, Encoding.UTF8);
                var parsed = JsonSerializer.Deserialize<AlarmData>(text, JsonOpts);
                return MergeWithDefaults(parsed);
            }
            catch
            {
                return DefaultData();
            }
        }

        /// <summary>原子写入：写到 .tmp，再重命名 + 自动备份</summary>
        public static void Save(AlarmData data, string? file = null)
        {
            file = file ?? DefaultFile;
            EnsureDir(file);
            string text = JsonSerializer.Serialize(data, JsonOpts);
            string tmp = file + ".tmp";
            File.WriteAllText(tmp, text, new UTF8Encoding(false));
            try
            {
                if (File.Exists(file))
                {
                    string bak = file + ".bak";
                    File.Copy(file, bak, true);
                }
            }
            catch { /* 备份失败不影响保存 */ }
            File.Move(tmp, file, true);
        }

        /// <summary>添加日志，保留最近 200 条</summary>
        public static void AppendLog(AlarmData data, TriggerLog entry)
        {
            if (data.Logs == null) data.Logs = new System.Collections.Generic.List<TriggerLog>();
            data.Logs.Add(entry);
            if (data.Logs.Count > 200)
            {
                data.Logs = data.Logs.Skip(data.Logs.Count - 200).Take(200).ToList();
            }
        }

        /// <summary>导出 JSON</summary>
        public static string ExportJson(AlarmData data) => JsonSerializer.Serialize(data, JsonOpts);

        /// <summary>导入 JSON</summary>
        public static AlarmData ImportJson(string text)
        {
            var parsed = JsonSerializer.Deserialize<AlarmData>(text, JsonOpts);
            return MergeWithDefaults(parsed);
        }

        /// <summary>生成新闹钟 ID</summary>
        public static string NewId() =>
            "a_" + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString("x") + "_" + Guid.NewGuid().ToString("N").Substring(0, 6);
    }
}
